lib/processor/download.js

```diff
@@ -0,0 +1,61 @@
+'use strict';
+
+const url = require('url');
+const path = require('path');
+const fs = require('fs');
+const Q = require('q');
+const http = require('follow-redirects').http;
+const https = require('follow-redirects').https;
+const s3 = new (require('aws-sdk')).S3();
+
+// get a remote file source
+module.exports = (uploadUrl) => {
+  let fileName = path.basename(url.parse(uploadUrl || 'unknown').pathname);
+  let writeStream = fs.createWriteStream(`/tmp/${fileName}`)
+  let result = {name: fileName, localPath: `/tmp/${fileName}`};
+
+  return Q.Promise((resolve, reject) => {
+    const rejectNoRetry = err => {
+      err.noRetry = true;
+      reject(err);
+    }
+
+    writeStream.on('error', err => reject(err));
+    writeStream.on('close', () => {
+      if (result.contentLength && +result.contentLength !== writeStream.bytesWritten) {
+        reject(new Error(`Expected ${result.contentLength} bytes from ${uploadUrl} - got ${writeStream.bytesWritten}`));
+      } else {
+        resolve(result);
+      }
+    });
+
+    if (!uploadUrl) {
+      rejectNoRetry(new Error('No url set for event'));
+    } else if (uploadUrl.startsWith('s3://')) {
+      let m = uploadUrl.match(/s3:\/\/([^\/]+)\/(.*)/);
+      result.s3Bucket = m[1];
+      result.s3Key = m[2];
+      s3.getObject({Bucket: m[1], Key: m[2]})
+        .createReadStream()
+        .on('error', err => rejectNoRetry(err))
+        .pipe(writeStream);
+    } else if (uploadUrl.match(/^https?:\/\//)) {
+      (uploadUrl.match(/^https/) ? https : http).get(uploadUrl, resp => {
+        if (resp.statusCode !== 200) {
+          let err = new Error(`Got ${resp.statusCode} for url: ${uploadUrl}`);
+          if (resp.statusCode < 500) {
```

Comment:
why < 500? [ticket](https://github.com/PRX/cms-audio-lambda/issues/8) just said retry on non-404s, no?

Reply:
So specifically, I want to retry 502/503/504 errors, as they indicate some sort of temporary problem upstream that _might_ resolve itself if I retry it.  (And experience has told me that S3 can periodically throw those).

Generally... any >= 500 is unexpected, and might change with a retry, whereas anything less than that won't.  But as I'm typing this, it's starting to sound like it might be better if I call out the statuses specifically.  Okay fine, you've talked me into it.

Reply:
glad to be of help here